File: MultiLanguageSandbox/src/test/deps/php/33.php
<?php

/**
    * Counts the number of flu infections in a dormitory grid after a certain number of days.
    * The dormitory is represented as an n*n grid where each cell can have:
    * - '.' representing a healthy person
    * - '#' representing an empty room
    * - '@' representing a person infected with the flu
    * Each day, infected individuals spread the flu to adjacent (up, down, left, right) healthy individuals.
    * The function calculates the total number of infected individuals after 'm' days.
    * Input: 
    *   - $dormitory: A 2D array representing the n*n dormitory grid.
    *   - $days: An integer representing the number of days.
    * Output: Returns the total number of flu infections after 'm' days.
    * Example:
    *   Input: [['.','.','.','.','#'], 
                ['.','#','.','@','.'], 
                ['.','#','@','.','.'], 
                ['#','.','.','.','.'], 
                ['#','.','.','.','.']], 4
    *   Output: 16
*/
function countFluInfections($dormitory, $days) {

    $n = count($dormitory);
    $infected = [];
    
    // Initialize the queue with initial infected positions
    for ($i = 0; $i < $n; $i++) {
        for ($j = 0; $j < $n; $j++) {
            if ($dormitory[$i][$j] == '@') {
                $infected[] = [$i, $j];
            }
        }
    }
    
    $directions = [[-1, 0], [1, 0], [0, -1], [0, 1]];
    
    for ($day = 0; $day < $days; $day++) {
        $newInfected = [];
        foreach ($infected as $cell) {
            list($x, $y) = $cell;
            foreach ($directions as $dir) {
                $nx = $x + $dir[0];
                $ny = $y + $dir[1];
                if ($nx >= 0 && $nx < $n && $ny >= 0 && $ny < $n && $dormitory[$nx][$ny] == '.') {
                    $dormitory[$nx][$ny] = '@';
                    $newInfected[] = [$nx, $ny];
                }
            }
        }
        $infected = $newInfected;
        if (empty($infected)) {
            break; // No more infections possible
        }
    }
    
    $total = 0;
    for ($i = 0; $i < $n; $i++) {
        for ($j = 0; $j < $n; $j++) {
            if ($dormitory[$i][$j] == '@') {
                $total++;
            }
        }
    }
    
    return $total;
}
function main(){
    assert(countFluInfections([
        ['.','.','.','.','#'], 
        ['.','#','.','@','.'], 
        ['.','#','@','.','.'], 
        ['#','.','.','.','.'], 
        ['#','.','.','.','.']], 4) === 16);
    

    assert(countFluInfections([
            ['.','.','.','.','#'], 
            ['.','#','.','@','.'], 
            ['.','#','@','.','.'], 
            ['#','.','.','.','.'], 
            ['#','.','.','.','.']], 1) === 2);
    

    assert(countFluInfections([
                ['.','.','.','.','#'], 
                ['.','#','.','@','.'], 
                ['.','#','@','.','.'], 
                ['#','.','.','.','.'], 
                ['#','.','.','.','.']], 2) === 7);

    assert(countFluInfections([
                    ['.','.','.','.','#'], 
                    ['.','#','.','#','.'], 
                    ['.','#','#','.','.'], 
                    ['#','.','.','.','.'], 
                    ['#','.','.','.','.']], 100) === 0);
}


main();
?>

File: MultiLanguageSandbox/src/test/deps/dart/33.dart

int calculateSavings(List<int> budgets)
/*
BB has always managed her pocket money by herself. At the beginning of each month, her mother gives BB 300 yuan, from which BB budgets her monthly expenses and consistently matches her actual expenses with her budget.
To teach BB how to save money, her mother proposed that BB could deposit any whole hundreds of yuan with her at any time, and at the end of the year, she would return the money to BB with an additional 20% interest. Therefore, BB devised a savings plan: at the beginning of each month, after receiving her pocket money from her mother, if she anticipates having more than 100 yuan or exactly 100 yuan left at the end of the month, she would deposit the whole hundreds with her mother and keep the remainder with herself.
BB realized that the main risk of this savings plan is that the money deposited with her mother cannot be withdrawn before the end of the year. It is possible that at the beginning of some month, the money BB has plus the money given by her mother for that month is not enough for the month's budget. If this situation arises, BB will have to economize and cut back on her budget for that month.
Now, according to BB's budget from January to December 2004, determine whether this situation will occur. If not, calculate how much money BB will have at the end of 2004 after her mother returns her savings with an additional 20% interest.

## Input Format

12 lines of data, representing BB's budget from January to December.

## Output Format

A single integer. If there is a month during the implementation of the savings plan where the money is insufficient, output -X, where X indicates the first month when this situation occurs; otherwise, output how much money BB will have at the end of 2004.

## example
calculateSavings([290, 230, 280, 200, 300, 170, 340, 50, 90, 80, 200, 60]) => -7
*/
{
  int currentMoney = 0;
  int savings = 0;
  
  for (int month = 0; month < 12; month++) {
    // Receive monthly allowance
    currentMoney += 300;
    
    // Check if budget can be met
    if (currentMoney < budgets[month]) {
      return -(month + 1); // Months are 1-based in output
    }
    
    // Apply budget
    currentMoney -= budgets[month];
    
    // Save whole hundreds if possible
    if (currentMoney >= 100) {
      int wholeHundreds = (currentMoney ~/ 100) * 100;
      savings += wholeHundreds;
      currentMoney -= wholeHundreds;
    }
  }
  
  // Calculate total after adding 20% interest to savings
  int totalSavingsWithInterest = (savings * 120) ~/ 100;
  int finalAmount = currentMoney + totalSavingsWithInterest;
  
  return finalAmount;
}
void main() {
  assert(calculateSavings([290, 230, 280, 200, 300, 170, 340, 50, 90, 80, 200, 60]) == -7);
  assert(calculateSavings([290, 230, 280, 200, 300, 170, 330, 50, 90, 80, 200, 60]) == 1580);
  assert(calculateSavings([300, 300, 300, 300, 300, 300, 300, 300, 300, 300, 300, 300]) == 0);
}